syntax = "proto3";

package simulate;

import "proto/encounter.proto";
import "proto/equipment.proto";
import "proto/talents.proto";
import "proto/policy.proto";

message SimulationConfig {
  // This represents the maximum number of iterations we're willing to perform.
  // Reasonable default: 10,000
  int32 max_iterations = 1;

  // We'll halt a simulation early if the error drops below this threshold.
  // Reasonable default: 0.01
  double error_target = 2;

  // We attack for time_target +/- time_variance.
  // Reasonable default: 5 minutes, or 300s (300000 ms)
  int32 time_target = 3;

  // We want combat to run about as long as the encounter says, but not exactly.
  // Reasonable default: 0.10, or 10%.
  double time_variance = 4;

  // `encounter_config` contains "how long" this fight is.
  EncounterConfig encounter_config = 5;

  // `equipment_config` contains the gearset we're evaluating our policy on.
  EquipmentConfig equipment_config = 6;

  // `talent_config` contains the set of talents we're evaluating our policy on.
  TalentConfig talent_config = 7;

  // `policy` represents the current policy we want our simulator to evaluate.
  // The model type may be either APL (deterministic) or DL (learned).
  PolicyConfig policy = 8;
}

message SimulationResult {
  double dps_mean = 1;
  double dps_variance = 2;
  string metadata = 3;
}
